### ISSUE TITLE
could not find capabilities for domaintype=qemu on mac

### ISSUE THREAD

**Issue by furlongm** [vagrant-libvirt/vagrant-libvirt#1017]
I _almost_ have vagrant-libvirt running on macos, but running into an issue.

vagrant, libvirt and qemu are installed from homebrew.
vagrant-libvirt is installed with `vagrant plugin install vagrant-libvirt`

Sample vagrant file:
```
Vagrant.configure(2) do |config|
  config.vm.provider :libvirt do |lv|
    lv.driver = 'qemu'
    lv.socket = '/usr/local/var/run/libvirt/libvirt-sock'
    lv.cpus = 1
    lv.memory = 512
  end
  config.vm.define 'test' do |node|
    node.vm.hostname = 'test'
    node.vm.box = 'generic/ubuntu1804'
  end
end
```

```
$ vagrant up
Bringing machine 'test' up with 'libvirt' provider...
==> test: Checking if box 'generic/ubuntu1804' version '1.9.16' is up to date...
==> test: Creating image (snapshot of base box volume).
==> test: Creating domain with the following settings...
==> test:  -- Name:              t_test
==> test:  -- Domain type:       qemu
==> test:  -- Cpus:              1
==> test:  -- Feature:           acpi
==> test:  -- Feature:           apic
==> test:  -- Feature:           pae
==> test:  -- Memory:            512M
==> test:  -- Management MAC:    
==> test:  -- Loader:            
==> test:  -- Nvram:             
==> test:  -- Base box:          generic/ubuntu1804
==> test:  -- Storage pool:      default
==> test:  -- Image:             /var/lib/libvirt/images/t_test.img (32G)
==> test:  -- Volume Cache:      default
==> test:  -- Kernel:            
==> test:  -- Initrd:            
==> test:  -- Graphics Type:     vnc
==> test:  -- Graphics Port:     -1
==> test:  -- Graphics IP:       127.0.0.1
==> test:  -- Graphics Password: Not defined
==> test:  -- Video Type:        cirrus
==> test:  -- Video VRAM:        256
==> test:  -- Sound Type:	
==> test:  -- Keymap:            en-us
==> test:  -- TPM Path:          
==> test:  -- INPUT:             type=mouse, bus=ps2
Error while creating domain: Error saving the server: Call to virDomainDefineXML failed: invalid argument: could not find capabilities for domaintype=qemu
```

**Comment by electrofelix** [comment]
Suggests that libvirt and qemu may not be working quite right.

Try running:
```
virsh capabilities | grep "domain type"
```
and see if it lists `qemu` in the output, as otherwise need to dig into what's needed to get libvirt & qemu to work together first

**Comment by furlongm** [comment]
```
$ virsh capabilities | grep "domain type"
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
      <domain type='qemu'/>
```

**Comment by electrofelix** [comment]
@furlongm are you able to create a vm using libvirt  directly?

**Comment by furlongm** [comment]
Aha - I can create using qemu:///session, but not qemu:///system

The above capabilities output is from session, but if I use system, I get the following:

```
$ sudo virsh capabilities | grep qemu
<baselabel type='qemu'>+819786158:+1133313938</baselabel>
```

**Comment by ladar** [comment]
@furlongm I haven't tried using libvirt/qemu on MacOS, but at least on Linux, you need to ensure the libvirt daemon is running, and that the current user is a member of the `kvm`, `qemu`, and the `libvirt` groups. Note the group names may be slightly different on MacOS. 

It's also possible to force a user session VM using the `Vagrantfile` but the last time I tried that path, it was a bit bumpy.

**Comment by murphye** [comment]
I am very interested in getting qemu and libvirt working on Mac as well. I have had some success but am currently running into an error.

`/Users/ermurphy/.vagrant.d/gems/2.4.6/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/driver.rb:121:in `list_all_networks': Call to virConnectListAllNetworks failed: this function is not supported by the connection driver: virConnectListAllNetworks (Libvirt::RetrieveError)`

I also had to create symbolic link
`sudo ln -s ~/.cache/libvirt/libvirt-sock /var/run/libvirt/libvirt-sock`

Current Vagrantfile

`Vagrant.configure(2) do |config|
  config.vm.provider :libvirt do |lv|
    lv.driver = 'qemu'
    lv.cpus = 1
    lv.memory = 512

    # Use QEMU session instead of system connection
    lv.qemu_use_session = true
    # URI of QEMU session connection, default is as below
    lv.uri = 'qemu:///session'
    # URI of QEMU system connection, use to obtain IP address for management
    # TODO: Unsure what to do here on Mac running in user space
    lv.system_uri = 'qemu:///system' 
    # Management network device
    #lv.management_network_device = 'virbr0'

  end

  config.vm.define 'test' do |node|
    node.vm.hostname = 'test'
    node.vm.box = 'centos/7'

    node.vm.network :public_network,
      :dev => "virbr1",
      :mode => "bridge",
      :type => "bridge"
  end
end`

Please let me know if there is anything obviously wrong here. I am new to Vagrant, Qemu, and Libvirt.

I am interested in collaborating to resolve this issue. I would like to demonstrate CentOS 7 running.

**Comment by ladar** [comment]
@murphye if I were to guess, that error is telling you that libvirt is looking for the network interface you specificed, aka `virbr1` but it can't be found, or created. If you type `ifconfig` do you see any `virbrXX` interfaces listed? You should... see the screenshot below. On my system, I think the `default` network aka `virbr0` was created when libvirt was installed, and `virbr1` was created at some point specifically for vagrant guests. That would mean the `vagrant-libvirt` plugin has to create its virtual interface on the host, which requires OS specific logic. I have no idea if the plugin code has that currently.

![Screenshot from 2019-07-10 17-02-49](https://user-images.githubusercontent.com/1434008/60966015-ebba6900-a334-11e9-9b4d-6d1a6de2a90c.png)

**Comment by ladar** [comment]
P.S. If you do have a default network, aka `virbr0` interface already, perhaps specifying it explicitly in your Vagrantfile will side step this issue.

**Comment by murphye** [comment]
@ladar Thanks for the deploy.

1) There is not any `virbrXX` interfaces listed when I do `ifconfig`.
2) If I play around with the device (i.e. try `bridge0` which I see in `ifconfig`, I get a different error:

```/Users/ermurphy/.vagrant.d/gems/2.4.6/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/driver.rb:59:in `open_read_only': Call to virConnectOpenReadOnly failed: internal error: unexpected QEMU URI path '/system', try qemu:///session (Libvirt::ConnectionError)```

I am way over my head here on diagnosing the problem since I am new to these tools. Would be wonderful if someone could create a guide for using QEMI/LibVirt/Vagrant on Mac OS.

**Comment by ladar** [comment]
@murphye that error implies it had trouble connecting to the libvirt system daemon, and it's suggesting you use a local libvirt session. Not sure how, or why a network config change would effect that.

The bottom line is that it appears the plugin doesn't support MacOS. It may be possible to workaround the implementation gaps by manually configuring the network, etc. But I can't say how difficult that will be.

What I think you need to try doing is "configuring" a virtual machine network, aka `virbrXX` interface. That virtual interface will use a driver that allows the guest to access it. If you have the `virt-manager` tool available on your platform, it has a network connection editor. That's where the GUI screenshot above came from. It might help you setup the guest network.

**Comment by furlongm** [comment]
Root cause seems to be running libvirtd as root on MacOS has permission issues when trying to change group permissions of qemu binaries. See https://github.com/Homebrew/homebrew-core/issues/41562

**Comment by Tcharl** [comment]
Hi,

Tried the same on macOS and failed too...

created a baked-in packer image with 
```
      "use_default_display": true,
      "display": "none",
      "accelerator": "hvf",
```
Was surprised it worked :-).

Installed libvirt via brew but started the service with sudo (so that I had the qemu://system available)
I also had to symlink libvirt-sock:
`sudo ln -s /usr/local/var/run/libvirt/libvirt-sock /var/run/libvirt/libvirt-sock`

Then installed vagrant-libvirt with that wonderful command:
`CONFIGURE_ARGS='with-ldflags=-L/opt/vagrant/embedded/lib with-libvirt-include=/usr/local/include/libvirt with-libvirt-lib=/usr/local/lib' GEM_HOME=~/.vagrant.d/gems GEM_PATH=$GEM_HOME:/opt/vagrant/embedded/gems PATH=/opt/vagrant/embedded/bin:$PATH vagrant plugin install vagrant-libvirt`

But the simplest vagrantfile fails badly:
```
ENV['VAGRANT_DEFAULT_PROVIDER'] = 'libvirt'
Vagrant.configure("2") do |config|
  config.vm.define "fedo-01" do |config|
      config.vm.hostname = "fedo-01"
      config.vm.box = "fedora-31-x86_64.libvirt.box"
      config.vm.box_check_update = false
      config.vm.synced_folder ".", "/vagrant", disabled: true # do not work otherwise, try to start some SMB magic
      config.vm.provider :libvirt do |v|
        v.memory = 1024
      end
  end
end
```
which outputs
`/Users/charliemordant/.vagrant.d/gems/2.4.9/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/util/network_util.rb:117:in `list_networks': Call to virConnectNumOfNetworks failed: this **function is not supported by the connection driver**: virConnectNumOfNetworks (Libvirt::RetrieveError)
`

This seems to be due to the fact that network is not valid on MacOS: https://github.com/libvirt/libvirt/blob/master/configure.ac#L198

How can we go further? I'm far from being a libvirt/qemu guy, just trying to make it work to have better perf than VirtualBox...

**Comment by shuxiao9058** [comment]
> Hi,
> 
> Tried the same on macOS and failed too...
> 
> created a baked-in packer image with
> 
> ```
>       "use_default_display": true,
>       "display": "none",
>       "accelerator": "hvf",
> ```
> 
> Was surprised it worked :-).
> 
> Installed libvirt via brew but started the service with sudo (so that I had the qemu://system available)
> I also had to symlink libvirt-sock:
> `sudo ln -s /usr/local/var/run/libvirt/libvirt-sock /var/run/libvirt/libvirt-sock`
> 
> Then installed vagrant-libvirt with that wonderful command:
> `CONFIGURE_ARGS='with-ldflags=-L/opt/vagrant/embedded/lib with-libvirt-include=/usr/local/include/libvirt with-libvirt-lib=/usr/local/lib' GEM_HOME=~/.vagrant.d/gems GEM_PATH=$GEM_HOME:/opt/vagrant/embedded/gems PATH=/opt/vagrant/embedded/bin:$PATH vagrant plugin install vagrant-libvirt`
> 
> But the simplest vagrantfile fails badly:
> 
> ```
> ENV['VAGRANT_DEFAULT_PROVIDER'] = 'libvirt'
> Vagrant.configure("2") do |config|
>   config.vm.define "fedo-01" do |config|
>       config.vm.hostname = "fedo-01"
>       config.vm.box = "fedora-31-x86_64.libvirt.box"
>       config.vm.box_check_update = false
>       config.vm.synced_folder ".", "/vagrant", disabled: true # do not work otherwise, try to start some SMB magic
>       config.vm.provider :libvirt do |v|
>         v.memory = 1024
>       end
>   end
> end
> ```
> 
> which outputs
> `/Users/charliemordant/.vagrant.d/gems/2.4.9/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/util/network_util.rb:117:in `list_networks': Call to virConnectNumOfNetworks failed: this **function is not supported by the connection driver**: virConnectNumOfNetworks (Libvirt::RetrieveError)
> `
> 
> This seems to be due to the fact that network is not valid on MacOS: https://github.com/libvirt/libvirt/blob/master/configure.ac#L198
> 
> How can we go further? I'm far from being a libvirt/qemu guy, just trying to make it work to have better perf than VirtualBox...

the same issue, do you have resolved this problem?

**Comment by radeksimko** [comment]
> This seems to be due to the fact that network is not valid on MacOS: https://github.com/libvirt/libvirt/blob/master/configure.ac#L198

I don't know enough about libvirt to tell what exactly does "network" driver in the comment refers to, but according to this compatibility table it appears that most APIs beyond the basic hypervisor ones just aren't available for QEMU at all (on any platform):
https://libvirt.org/hvsupport.html#virNWFilterDriver

You can in fact reproduce this outside of Vagrant:

```
$ virsh -c qemu:///system?socket=/usr/local/var/run/libvirt/libvirt-sock net-list --all
error: Failed to get the number of active networks
error: this function is not supported by the connection driver: virConnectNumOfNetworks
```

Which makes me think that this Vagrant plugin won't work in most scenarios which require any network configuration in QEMU (and need these libvirt API calls). There isn't much that can be done on the plugin side, perhaps other than calling QEMU directly or implementing these APIs upstream in libvirt.

On that note I feel the best course of action for these scenarios may be creation of a dedicated QEMU Vagrant plugin. This is how Packer integrates with QEMU too:
https://www.packer.io/docs/builders/qemu.html
https://github.com/hashicorp/packer/blob/cc367bd0b7314c8810d6228c5f97e93950c59ca8/builder/qemu/driver.go#L79-L80
this would avoid any libvirt API limitations and made any QEMU boxes coming out of Packer compatible with such Vagrant provider.

**Comment by Tcharl** [comment]
As a backend developper, I always preferred to use Java rather than Assembly, so I would prefer a contribution to libvirt (that won't be only useful for this plugin). However, I'm just discovering Qemu and infra/VM unix ecosystem so for sure not a good source of truth...

**Comment by s-alty** [comment]
Here's a related issue that was filed against the libvirt project:
https://gitlab.com/libvirt/libvirt/-/issues/75

**Comment by electrofelix** [comment]
@radeksimko the idea of a direct qemu driver would be interesting, I did in the past move some code to a separate driver https://github.com/vagrant-libvirt/vagrant-libvirt/blob/master/lib/vagrant-libvirt/driver.rb to hide some of the details from the rest of the code for parts that might be more specific to libvirt such as the connection setup.

Possibly the same could be done for libvirt parts that may not work on MacOS so the code could call different APIs depending on the OS

**Comment by ladar** [comment]
@electrofelix this is an old thread, so I won't dive deep. Just let me know if there are any changes (hopefully trivial) that need to be made to the robox box images, so they work properly using this hypothetical `qemu` driver.

I don't believe there is anything, in the image itself that would need changing but I could be wrong. Areas of concern though, are the fact that `kvm` is setup as the accelerator via the embedded Vagrantfile, so perhaps additional logic needs to be added there to support things like `hvf` etc. 

Of course if it becomes a full fledged `vagrant` plugin, and if `packer` can generate Vagrant boxes for it, I don't mind adding them to the robox build pipeline. 

I occasionally look at the list of providers on [vagrantcloud.com](https://vagrantcloud.com), and the platforms that `packer`  supports for anything that might be worth supporting. But it seems the focus these days is on cloud providers, not locally hosted boxes. That said I noticed `qemu` provider, along `lxd` and `lxd` and have kicked around the idea of adding them. But those platforms only have a handful of boxes published, and relatively few downloads, so which made me think they aren't popular enough to bother supporting just yet.

**Comment by electrofelix** [comment]
@ladar I don't think the problem is with the boxes, it's more there are issues with libvirt running on MacOS, and then a couple of issues with vagrant-libvirt.

qemu:///session support is now better in vagrant-libvirt, there is one bug that needs to be fixed #986 to allow additional disks, and there is an issue in how we currently pick up `LIBVIRT_DEFAULT_URI` env var. So both are fixed, there should be less issues getting to the point of starting the VM.

However I think the core of this issue is libvirt network support on MacOS. If it's possible to bring up a machine on MacOS using virt-manager, we should be able to bring it up with Vagrant. However it's unclear to me what the network should look like, and if we can't use certain APIs we'll need to know what the alternatives should be.

If someone with MacOS is available to experiment with virt-manager and libvirt and attach some of the XML for both the VM and network if possible to configure a network (maybe it's only static is possible, or requires a bridge manually created, etc), then it should be possible to work out what vagrant-libvirt needs to do to map behaviours, or at least confirm if it's possible.

**Comment by ladar** [comment]
@electrofelix you'll find there are a lot small, but critical differences with MacOS. It's based off BSD, so many of the tools, commands, and APIs are very different from what you'd find on a Linux system. I've had to make a number of tweaks to my Robox scripts so they work on MacOS. When you also mix in a the different accelerator, your best to think of it as an entirely new platform, that is just libvirt/qemu in name only. To give you an analogy, it's like moving from the iPhone to a Samsung Galaxy. They're both smart phones, but they are completely different under the hood. 

As for session support, I gave it a try last night, and I'm happy to say it worked rather well, save one issue. I tried to configure where I wanted the disk image file to go, and instead of placing where I wanted, the disk went to $HOME/.local/share/libvirt/images/ ... they default storage pool location for user sessions. I'm glad I checked or and aborted the provisioning, or I'd have generated a couple tereabytes of disk writes on my SSDs. I was compiling Lineage, and generating firmware update disks for servers, and didn't want all that IO to wear out my SSDs. Hence a user session, with the writes going to an external USB HD. Perhaps I'm expecting too much, but wouldn't it make more sense for Vagrant to create those pools (if they're required), and then dispose of them when the box is destroyed? Vagrant does something like that with Hyper-V. It will create a virtual switch if one isn't configured already. And if it creates it, it will destroy it when finished. But if it exists already, it will leave it alone. 

To finish my story, as it turns out, once I manually created the storage pools at each the location, everything worked as expected. 

As for the MacOS issue. I have a 2011 MacBook Air that was used to build the Parallels VMs once upon a time. That is until the SSD literally died. I need to order a replacement from China (my first replacement got stolen out of my mailbox, and I need the special screw drive to open it up). 

Anyways, if your determined to make it work, I could set that up, and you could SSH in and use it to experiment. I've moved all Parallels box builds to a slightly newer (2015) Mac Mini, so that MacBook Air is essentially surplus. Let me know if your interested and I'll dig it out of the drawer and work on getting it fixed.

**Comment by electrofelix** [comment]
@ladar I would love to get to the bottom on at least understanding what is feasible, so either we can say this can't work at all, or it can only work provided you set up networking outside of vagrant-libvirt because libvirt can't do the network setup on macos. I don't know how much effort can be put towards it, I think mostly it would have to start with a request for someone to confirm that libvirt + virt-manager is usable on MacOS and that you can create and boot a VM and share the XML files generated.
The next step would be to confirm whether it's possible to have some form of network connectivity and  share either what was done in virt-manager and the resulting XML, and/or what had to be set up outside of libvirt and then passed in to allow a network connection.

I think the idea on vagrant removing stuff once it's finished with them would be good, possibly might need use of some comment tags embedded in the XML so that it's possible to establish that what was created by vagrant coupled with being able to establish that the pools are no longer needed which might be a little harder with qemu:///system than with qemu:///session.

**Comment by ladar** [comment]
@electrofelix like I said, if you're interested in getting it vagrant+libvirt working on Macs, I don't mind setting up access to the MacBook Air I mentioned. I can set you up with VNC/SSH access, once I get it repaired. It'll be a fresh install, so you can do/install whatever you need. I'll just wipe it when you're done. 

I (re) ordered the replacement HDD, and screwdriver, last week. And I checked on it today. Laughably, it won't get delivered until Dec 28th. Let me know if you're interested, and I can connect with you off-GH to get it setup.

![Screenshot from 2020-12-01 14-00-04](https://user-images.githubusercontent.com/1434008/100790789-f4a25380-33dd-11eb-8b49-124e364d202b.png)

**Comment by ladar** [comment]
@electrofelix regarding the network/storage volume setup for user session virtual machines... I think outputting a warning is probably what's important. Arguably, you could ask for permission with a y/n prompt as well. Although the latter would break automation. And since `vagrant` is basically a tool designed to automate things, I would personally vote against a prompt.

I mentioned Hyper-V above, but I mistakenly referred to `vagrant`. I should have said `packer` ... it does something along the lines I'm describing with its network setup. If `packer` detects an existing virtual switch (which must be connected to one of the system network interfaces), it will print a notice like the one below. Note the lines which mention a `Switch`. If `packer` doesn't find a suitable, existing virtual switch (ie if the existing virtual switches are setup as host only), then it will create a new virtual switch, which it then destroys once the build is finished.

Along these lines, I think `vagrant-libvirt` should do something similar if detects a path in the `Vagrantfile` that doesn't match a storage volume. The same goes if it needs to setup a new virtual network. Naturally, if it does create a storage volume/network along these lines, then it should, probably, delete** them during cleanup (aka when `vagrant destroy` is run). 

At the very least, in the short term, the plugin should detect a `Vagrantfile` which dictates a storage path that doesn't match an existing storage pool, and abort, rather than use the default storage pool. The way it works now is a little dangerous.

**Arguably, you could leave behind a virtual network setup. Historically `vagrant` creates its own bridge interface, if needed.

```
C:\Users\Administrator\Desktop\robox>packer build developer-hyperv.json
magma-developer-hyperv: output will be in this color.

==> magma-developer-hyperv: Creating build directory...
==> magma-developer-hyperv: Retrieving ISO
==> magma-developer-hyperv: Trying https://mirrors.edge.kernel.org/centos/6.10/isos/x86_64/CentOS-6.10-x86_64-minimal.iso
==> magma-developer-hyperv: Trying https://mirrors.edge.kernel.org/centos/6.10/isos/x86_64/CentOS-6.10-x86_64-minimal.iso?checksum=sha256%3A7c0dee2a0494dabd84809b72ddb4b761f9ef92b78a506aef709b531c54d30770
==> magma-developer-hyperv: https://mirrors.edge.kernel.org/centos/6.10/isos/x86_64/CentOS-6.10-x86_64-minimal.iso?checksum=sha256%3A7c0dee2a0494dabd84809b72ddb4b761f9ef92b78a506aef709b531c54d30770 => C:\Users\Administrator\Desktop\robox\packer_cache\10f0576b6e0a81ff333bdc21e56199174aed0925.iso
==> magma-developer-hyperv: Starting HTTP server on port 8931
==> magma-developer-hyperv: Creating switch 'Swich' if required...
==> magma-developer-hyperv:     switch 'Swich' already exists. Will not delete on cleanup...
==> magma-developer-hyperv: Creating virtual machine...
==> magma-developer-hyperv: Enabling Integration Service...
==> magma-developer-hyperv: Mounting os dvd drive C:\Users\Administrator\Desktop\robox\packer_cache\10f0576b6e0a81ff333bdc21e56199174aed0925.iso ...
==> magma-developer-hyperv: Skipping mounting Integration Services Setup Disk...
==> magma-developer-hyperv: Mounting secondary DVD images...
==> magma-developer-hyperv: Configuring vlan...
==> magma-developer-hyperv: Determine Host IP for HyperV machine...
==> magma-developer-hyperv: Host IP for the HyperV machine: 10.10.10.203
==> magma-developer-hyperv: Starting the virtual machine...
==> magma-developer-hyperv: Waiting 20s for boot...
==> magma-developer-hyperv: Typing the boot command...
==> magma-developer-hyperv: Waiting for SSH to become available...
==> magma-developer-hyperv: Connected to SSH!
```

![Screenshot from 2020-12-01 14-33-16](https://user-images.githubusercontent.com/1434008/100793809-4fd64500-33e2-11eb-9419-dd39f673cbb1.png)

**Comment by kenji21** [comment]
> @ladar I would love to get to the bottom on at least understanding what is feasible, so either we can say this can't work at all, or it can only work provided you set up networking outside of vagrant-libvirt because libvirt can't do the network setup on macos. I don't know how much effort can be put towards it, I think mostly it would have to start with a request for someone to confirm that libvirt + virt-manager is usable on MacOS and that you can create and boot a VM and share the XML files generated.

More a workaround than, but can we customise QEMU options ? The idea is to pass some option like these ones:
```
        <qemu:arg value='-netdev'/>
        <qemu:arg value='user,id=n1,hostfwd=tcp::2222-:22'/>
        <qemu:arg value='-device'/>
        <qemu:arg value='virtio-net-pci,netdev=n1,bus=pcie.0,addr=0x19'/>
```
Gathered from https://www.naut.ca/blog/2020/08/26/ubuntu-vm-on-macos-with-libvirt-qemu/, full XML template for Qemu is here: https://raw.githubusercontent.com/yoonsikp/vm_configs/master/ubuntu.xml

This way we can `vagrant up` and then ssh into the VM... which can be useful

**Comment by imperialguy** [comment]
Have the same issue on my Mac
```
/Users/testuser/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.3.0/lib/vagrant-libvirt/driver.rb:59:in `open_read_only': Call to virConnectOpenReadOnly failed: Failed to connect socket to '/usr/local/var/run/libvirt/libvirt-sock-ro': No such file or directory (Libvirt::ConnectionError)
```

Also, I explicitly specify the socket in my VagrantFile:
`v.socket = "/Users/testuser/.cache/libvirt/libvirt-sock"`

The same error occurs with either one of the drivers`kvm` or `qemu` (with or without session enabled)

**Comment by dcode** [comment]
So this is an old issue, but here we are...  *bump*

I managed to get `vagrant-libvirt` plugin installed using the [move iconv](https://github.com/vagrant-libvirt/vagrant-libvirt/issues/818#issuecomment-615802249) trick and paired with the homebrew installed version of `libvirt` (7.5.0) and `qemu` (6.0.0). I had to enable the unix socket by adding the following line to `/usr/local/etc/libvirt/libvirtd.conf`.

```unix_sock_dir = "/usr/local/var/run/libvirt"```

then run ```brew services restart libvirt```

After all that, I used this Vagrantfile:

```
Vagrant.configure("2") do |config|
  config.vm.box = "generic/centos8"
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = 'qemu'
    # Use QEMU session instead of system connection
    libvirt.qemu_use_session = true
  end
end
```

And we're to the point of failing listing networks (as noted [here](https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1017#issuecomment-608002209)).

Expections/Wishlist:

1. I would have hoped to use `hvf` as an accelerator vs plain emulated qemu (which is drastically slower).
2. Packer uses the user mode networking like @kenji21 mentioned in his work around, that seems like fairly simple since it's Qemu specific and not OS-specific. This is also consistent with the default vagrant NAT usage with VirtualBox. For example, you forward ports to services that you want to access, all VMs get the IP of `10.0.2.15` by default, etc.

Meanwhile, does anyone have a working configuration that could allow a user to manually create a bridge and use that?

**Comment by atreides322** [comment]
I was able to create a Vagrantfile that enables HVF and adds user-mode networking:
```ruby
Vagrant.configure('2') do |config|
  config.vm.synced_folder '.', '/vagrant', disabled: true
  config.vm.box = 'generic/ubuntu2004'
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = 'qemu'
    libvirt.uri = 'qemu:///session'
    libvirt.socket = '#{Dir.home}/.cache/libvirt/libvirt-sock'
    libvirt.qemuargs :value => '-machine'
    libvirt.qemuargs :value => 'accel=hvf'
    libvirt.qemuargs :value => '-netdev'
    libvirt.qemuargs :value => 'user,id=n0,hostfwd=tcp::2222-:22'
    libvirt.qemuargs :value => '-device'
    libvirt.qemuargs :value => 'virtio-net-pci,netdev=n0'
  end
end
```
I still get the same error about `Call to virConnectNumOfNetworks failed`:
```
/Users/atreides/.vagrant.d/gems/2.7.4/gems/vagrant-libvirt-0.7.0/lib/vagrant-libvirt/util/network_util.rb:144:in `list_networks': Call to virConnectNumOfNetworks failed: this function is not supported by the connection driver: virConnectNumOfNetworks (Libvirt::RetrieveError)
```
However, the domain does get defined. I am able to use `virsh` and manually start the domain. It boots up and I am able to use user/pass authentication to ssh into the vm with `ssh vagrant@localhost -p 2222` and the VM does have network access. Also, I was able to verify that the HVF acceleration is working because there is a distinct difference in performance between including and omitting `-machine accel=hvf`.

It seems like this is just hung up on the network driver explicitly disabled on macOS in libvirt.

**Comment by electrofelix** [comment]
@atreides322 that seems like a decent amount of progress. I think the remaining issue would be around avoiding the requirement to have a vagrant-libvirt management network, and it looks like you've provided a possible path to that.

I wasn't aware of the following setting:
```
    libvirt.qemuargs :value => '-netdev'
    libvirt.qemuargs :value => 'user,id=n0,hostfwd=tcp::2222-:22'
    libvirt.qemuargs :value => '-device'
    libvirt.qemuargs :value => 'virtio-net-pci,netdev=n0'
```

It seems likely that could be injected as an alternative to the existing port forward that is using the ssh tunnelling as well as for any other defined port forwards, and once the VM is accessible via that, it's successfully booted. Would certainly make it closer in behaviour to the VirtualBox provider, and it looks like it would solve another long standing issue in that the hostfwd setting appears to support udp as well as tcp.

It looks like the main reason it hasn't been used before is I don't see it exposed via libvirt which is why it would be necessary to pass through via qemuargs explicitly. However I don't see any down side to doing that and dropping the need for the management interface which would remove the need to list network devices.

It wouldn't solve the problem that networking isn't supported on MacOS, https://gitlab.com/libvirt/libvirt/-/issues/75 covers that, but it would allow simple Vagrantfile which don't need to add additional networks to work?

**Comment by atreides322** [comment]
@electrofelix I took inspiration from @kenji21 for adding the user-mode networking. It is also very similar to how Linux-on-Mac handles networking with Qemu. https://github.com/lima-vm/lima/blob/f42606966930d7675690990349dea0428dafbe7f/pkg/qemu/qemu.go#L271-L310

As far as port fowarding, it looks like Linux-on-Mac also uses SSH port forwarding. Presumably so that you don't have to bring down the VM just to change the port forwarding: https://github.com/lima-vm/lima/blob/f42606966930d7675690990349dea0428dafbe7f/pkg/hostagent/port.go

Even with manually defining the network adapter and manually starting the domain, I am unable to use any `vagrant` commands, such as `vagrant ssh`, `vagrant provision`, etc. It seems like it is trying to use `virtqemud-sock-ro`, which doesn't exist. Again, this is probably lack of network driver functionality, as mentioned.

```
/Users/atreides/.vagrant.d/gems/2.7.4/gems/vagrant-libvirt-0.7.0/lib/vagrant-libvirt/driver.rb:64:in `open_read_only': Call to virConnectOpenReadOnly failed: Failed to connect socket to '/usr/local/var/run/libvirt/virtqemud-sock-ro': No such file or directory (Libvirt::ConnectionError)
```

**Comment by petracvv** [comment]
I'm not sure it trying to use the `virtqemud-sock-ro` is due to the lack of network driver support. For some reason it is trying to use the modular qemu socket (https://libvirt.org/daemons.html#modular-sockets), instead of the monolithic one. The modular sockets don't work on MacOS as they appear to require DBUS running when I tried to launch the virtqemed daemon.

**Comment by alisson276** [comment]
@petracvv , this error of `virtqemud-sock-ro` can be solved by setting `v.uri = "qemu:///session?socket=/Users/USERNAME/.cache/libvirt/libvirt-sock"`, anyway, the network errror starts to happen again.

```
/Users/USERNAME/.vagrant.d/gems/2.7.4/gems/vagrant-libvirt-0.7.0/lib/vagrant-libvirt/util/network_util.rb:144:in `list_networks': Call to virConnectNumOfNetworks failed: this function is not supported by the connection driver: virConnectNumOfNetworks (Libvirt::RetrieveError)
```

**Comment by itsnavee** [comment]
starting libvirt with sudo solved it for me
```
sudo brew services restart libvirt
virt-manager -c qemu:///session --no-fork
```

**Comment by rustydb** [comment]
After using `sudo brew services restart libvirt` I hit this on macOS:

```bash
/Users/rusty/.vagrant.d/gems/2.7.6/gems/fog-libvirt-0.9.0/lib/fog/libvirt/requests/compute/vm_action.rb:7:in `create': Call to virDomainCreateWithFlags failed: '/usr/libexec/qemu-bridge-helper' is not a suitable bridge helper: No such file or directory (Libvirt::Error)
```

`/usr/libexec/qemu-bridge-helper` doesn't exist.

If I do this I see nothing:
```bash
virsh net-list --all
 Name   State   Autostart   Persistent
----------------------------------------
```

If I do this I see default, but I can't start it:
```bash
sudo virsh net-list --all
 Name      State      Autostart   Persistent
----------------------------------------------
 default   inactive   no          yes
```
```bash
sudo virsh net-start default
error: Failed to start network default
error: Cannot rename interface 'bridge1' to 'virbr0' on this platform: Function not implemented
```

**Comment by electrofelix** [comment]
@rustydb looks like at least there has been progress with libvirt to support networking. When using the session connection, it will cause libvirt to try and  I'd guess that the default network definition may not be correct though for the default network stack.

Can you try the following (`-c qemu:///system` should be the equivalent of using sudo to run virsh):
```
virsh -c qemu:///system net-dumpxml default
```

My guess is that `virbr0` is referenced in the XML definition of the network while the mac address referenced in the network definition matches `bridge1` on the host? You can try `virsh -c qemu:///system net-edit default` and adjust to see if this will work, but I'm guessing it'll still fail at the qemu-bridge-helper step.

Can probably see further up https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1017#issuecomment-974788480 that there is a possible path to removing the requirement to have a functioning network to get access to VMs spun up using libvirt, but there's just still a fair amount of work before I can get to it.

**Comment by rustydb** [comment]
I have to run it as `sudo`, but I get this:
```bash
[639]rusty@C02F43JEMD6P:~/scratch> virsh -c qemu:///system net-dumpxml default
error: failed to connect to the hypervisor
error: Failed to connect socket to '/usr/local/var/run/libvirt/libvirt-sock': Permission denied
[640]rusty@C02F43JEMD6P:~/scratch> sudo !!
sudo virsh -c qemu:///system net-dumpxml default
<network>
  <name>default</name>
  <uuid>8d134c8a-b980-4f5c-bc1b-ae98fe6d4388</uuid>
  <forward mode='nat'/>
  <bridge name='virbr0' stp='on' delay='0'/>
  <mac address='52:54:00:4a:7a:b9'/>
  <ip address='192.168.122.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.122.2' end='192.168.122.254'/>
    </dhcp>
  </ip>
</network>
```

**Comment by rustydb** [comment]
If I use the args from [comment](https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1017#issuecomment-974788480) I still hit the bridge helper issue.

This is the closest I've gotten, is by running libvirt with `sudo` and using `qemu:///session`. I can't use `qemu:///system` unless I'm root. I don't want to run Vagrant as root though, I'm trying to get a guide up for internal use at work.

**Comment by electrofelix** [comment]
Did you try using `bridge1` as the bridge name rather than `virbrX`?

To avoid running as root, need to look at the group owning `/usr/local/var/run/libvirt/libvirt-sock` and add your user to that group. On Ubuntu it's usually set up that `libvirt` is the group users need to be added to, to allow access to libvirt system context.
For Mac, might need to look at what is set for `unix_sock_group = "libvirt"` in the `/etc/libvirt/libvirtd.conf` file.

Since a lot of this is hooked into the rest of the distro config on Linux distros I'm not sure exactly what knobs need to be tweaked in the libvirtd.conf around the access controls. Hopefully it's already using a group for the socket file and you just need to add yourself to that group and then log out and back in.

**Comment by rustydb** [comment]
It'll make a new bridge interface and increment the number
```bash
[652]rusty@C02F43JEMD6P:~/scratch> virsh  ./default.xml
[652]rusty@C02F43JEMD6P:~/scratch> sudo virsh net-start default
error: Failed to start network default
error: Cannot rename interface 'bridge5' to 'bridge1' on this platform: Function not implemented

[653]rusty@C02F43JEMD6P:~/scratch> sudo virsh net-dumpxml default
<network>
  <name>default</name>
  <uuid>8d134c8a-b980-4f5c-bc1b-ae98fe6d4388</uuid>
  <forward mode='nat'/>
  <bridge name='bridge1' stp='on' delay='0'/>
  <mac address='52:54:00:4a:7a:b9'/>
  <ip address='192.168.122.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.122.2' end='192.168.122.254'/>
    </dhcp>
  </ip>
</network>
```

**Comment by rustydb** [comment]
The whole config file is commented out, so nothing is explicitly set. Here's the `unix_sock_group`:
```bash
sudo grep unix_sock_group /usr/local/etc/libvirt/libvirtd.conf
#unix_sock_group = "libvirt"
```

I should mention I'm on an x86_64 Mac (not ARM).

**Comment by electrofelix** [comment]
Does the `libvirt` group exist and is it the group for the socket file `/usr/local/var/run/libvirt/libvirt-sock` or what are the owner/group assigned to the file?

I'm not familiar enough with how the bridge devices should work on Mac, but perhaps try changing the mac address in the network device XML as well to match the bridge device that was already created. e.g. `bridge1` try using `8a:66:5a:c0:eb:01`.

**Comment by rustydb** [comment]
No dice when using the same MAC. In-fact every time I re-run `net-start` it makes a new NIC :) 

```
[667]rusty@C02F43JEMD6P:~/scratch> sudo virsh net-start default
error: Failed to start network default
error: Cannot rename interface 'bridge7' to 'bridge1' on this platform: Function not implemented

[668]rusty@C02F43JEMD6P:~/scratch> sudo virsh net-start default
error: Failed to start network default
error: Cannot rename interface 'bridge8' to 'bridge1' on this platform: Function not implemented

[669]rusty@C02F43JEMD6P:~/scratch> sudo virsh net-start default
error: Failed to start network default
error: Cannot rename interface 'bridge9' to 'bridge1' on this platform: Function not implemented
```
Perms/ownership on that file:
```bash
ll /usr/local/var/run/libvirt/libvirt-sock
srwx------ 1 root 0 Aug 31 12:03 /usr/local/var/run/libvirt/libvirt-sock
```

Nothing for libvirt in `/etc/group`, this might be because libvirt is installed via `brew` which wasn't ran (and I don't think it can be) as root.

```
674]rusty@C02F43JEMD6P:~/scratch> sudo cat /etc/group
##
# Group Database
#
# Note that this file is consulted directly only when the system is running
# in single-user mode.  At other times this information is provided by
# Open Directory.
#
# See the opendirectoryd(8) man page for additional information about
# Open Directory.
##
nobody:*:-2:
nogroup:*:-1:
wheel:*:0:root
daemon:*:1:root
kmem:*:2:root
sys:*:3:root
tty:*:4:root
operator:*:5:root
mail:*:6:_teamsserver
bin:*:7:
procview:*:8:root
procmod:*:9:root
owner:*:10:
everyone:*:12:
_taskgated:*:13:_taskgated
group:*:16:
staff:*:20:root
_networkd:*:24:
_installassistant:*:25:
_lp:*:26:
_postfix:*:27:
_postdrop:*:28:
certusers:*:29:root,_jabber,_postfix,_cyrus,_calendar,_dovecot
_keytabusers:*:30:_calendar,_jabber,_postfix
_scsd:*:31:
_ces:*:32:
_appstore:*:33:_appstore
utmp:*:45:
authedusers:*:50:
interactusers:*:51:
netusers:*:52:
consoleusers:*:53:
_mcxalr:*:54:
_appleevents:*:55:
_geod:*:56:
_devdocs:*:59:
_sandbox:*:60:
localaccounts:*:61:
netaccounts:*:62:
_mdnsresponder:*:65:
_uucp:*:66:
_ard:*:67:
dialer:*:68:
network:*:69:
_www:*:70:_devicemgr,_teamsserver
_eppc:*:71:_eppc
_cvs:*:72:
_svn:*:73:
_mysql:*:74:
_sshd:*:75:
_qtss:*:76:
_mailman:*:78:
_appserverusr:*:79:
admin:*:80:root
_appserveradm:*:81:
_clamav:*:82:
_amavisd:*:83:
_jabber:*:84:
_appowner:*:87:
_windowserver:*:88:
_spotlight:*:89:
accessibility:*:90:
_tokend:*:91:
_securityagent:*:92:
_calendar:*:93:_teamsserver
_teamsserver:*:94:_devicemgr
_update_sharing:*:95:
_installer:*:96:
_atsserver:*:97:
_lpadmin:*:98:
_unknown:*:99:
_lpoperator:*:100:
_softwareupdate:*:200:_softwareupdate
_guest:*:201:
_coreaudiod:*:202:
_screensaver:*:203:
_developer:*:204:
_locationd:*:205:
_detachedsig:*:207:_locationd
_trustevaluationagent:*:208:
_odchpass:*:209:_teamsserver
_timezone:*:210:
_lda:*:211:
_cvms:*:212:
_usbmuxd:*:213:
_postgres:*:216:_devicemgr,_calendar,_teamsserver,_xserverdocs
_devicemgr:*:220:
_webauthserver:*:221:_teamsserver,_devicemgr
_netbios:*:222:
_warmd:*:224:_warmd
_dovenull:*:227:
_netstatistics:*:228:
_assetcache:*:235:
_coremediaiod:*:236:
_launchservicesd:*:239:
_iconservices:*:240:
_distnote:*:241:
_nsurlsessiond:*:242:
_displaypolicyd:*:244:
_astris:*:245:
_gamecontrollerd:*:247:
_mbsetupuser:*:248:
_ondemand:*:249:
_analyticsusers:*:250:_analyticsd,_networkd,_timed,_reportmemoryexception,_mbsetupuser
_xserverdocs:*:251:
_wwwproxy:*:252:
_mobileasset:*:253:
_findmydevice:*:254:
_datadetectors:*:257:
_captiveagent:*:258:
_ctkd:*:259:
_applepay:*:260:
_hidd:*:261:
_cmiodalassistants:*:262:
_analyticsd:*:263:_analyticsd
_webdeveloper:*:264:
_fpsd:*:265:_fpsd
_timed:*:266:
_nearbyd:*:268:_nearbyd
_reportmemoryexception:*:269:_reportmemoryexception
_driverkit:*:270:_driverkit
_diskimagesiod:*:271:_diskimagesiod
_logd:*:272:_logd
_appinstalld:*:273:_appinstalld
_installcoordinationd:*:274:_installcoordinationd
_demod:*:275:_demod
_rmd:*:277:_rmd
_accessoryupdater:*:278:_accessoryupdater
_knowledgegraphd:*:279:_knowledgegraphd
_coreml:*:280:_coreml
_sntpd:*:281:_sntpd,_timed
_trustd:*:282:_trustd
_darwindaemon:*:284:_darwindaemon
_notification_proxy:*:285:
com.apple.access_ftp:*:395:
com.apple.access_disabled:*:396:
com.apple.access_sessionkey:*:397:
com.apple.access_screensharing:*:398:
com.apple.access_ssh:*:399:
com.apple.access_remote_ae:*:400:
_oahd:*:441:_oahd
```

**Comment by rustydb** [comment]
@electrofelix I am following up on this issue in these threads (vs. hijacking this one):
- https://gitlab.com/libvirt/libvirt/-/issues/275
- https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1449
